helix-lsp/src/snippet.rs

```diff
@@ -0,0 +1,518 @@
+                // TODO: variables. For now, fall back to the default, which defaults to "".
+                let text = r#default.unwrap_or_default();
+                offset += text.chars().count();
+                insert.push_str(text);
+            }
+            Tabstop { tabstop } => {
+                tabstops.push((tabstop, offset, offset));
+            }
+            Placeholder { tabstop, value } => match value.as_ref() {
+                // https://doc.rust-lang.org/beta/unstable-book/language-features/box-patterns.html
+                // would make this a bit nicer
+                Text(text) => {
+                    if include_placeholer {
+                        let len_chars = text.chars().count();
+                        tabstops.push((tabstop, offset, offset + len_chars + 1));
+                        offset += len_chars;
+                        insert.push_str(text);
+                    } else {
+                        tabstops.push((tabstop, offset, offset));
+                    }
+                }
+                other => {
+                    log::error!(
+                        "Discarding snippet: generating a transaction for placeholder contents {:?} is unimplemented.",
+                        other
+                    );
+                    return Transaction::new(doc);
+                }
+            },
+            other => {
+                log::error!(
+                    "Discarding snippet: generating a transaction for {:?} is unimplemented.",
+                    other
+                );
+                return Transaction::new(doc);
+            }
+        }
+    }
+
+    let transaction = Transaction::change_by_selection(doc, selection, |range| {
+        let cursor = range.cursor(text);
+        (
+            (cursor as i128 + start_offset) as usize,
+            (cursor as i128 + end_offset) as usize,
+            Some(insert.clone().into()),
+        )
+    });
+
+    // sort in assending order (except for 0, which should always be the last one (per lsp doc))
+    tabstops.sort_unstable_by_key(|(n, _o1, _o2)| if *n == 0 { usize::MAX } else { *n });
+
+    // merge tabstops with the same index (we take advantage of the fact that we just sorted them
+    // above to simply look backwards)
+    let mut ntabstops = Vec::<SmallVec<[(usize, usize); 1]>>::new();
+    {
+        let mut prev = None;
+        for (tabstop, o1, o2) in tabstops {
+            if prev == Some(tabstop) {
+                let len_1 = ntabstops.len() - 1;
+                ntabstops[len_1].push((o1, o2));
+            } else {
+                prev = Some(tabstop);
+                ntabstops.push(smallvec![(o1, o2)]);
+            }
+        }
+    }
+
+    if let Some(first) = ntabstops.first() {
+        let cursor_offset = insert.chars().count() as i128 - (end_offset - start_offset);
+        let mut extra_offset = start_offset;
+        transaction.with_selection(selection.clone().transform_iter(|range| {
```

Comment:
It may be easier to update selection based on the transaction first (`selection.clone().map(transaction.changes())`). That way you don't have to worry about maintaining extra offset and just compute the desired position based on each selection cursor and the tabstop range offset.